File: fortran/fizzbuzz.f90
module simple_fizzbuzz_mod
  ! Reduces code reuse at the expense of overall length.
  ! Slight gain in flexibility?
  implicit none

  contains

  function divisible_by(num, denom)
    integer, intent(in) :: num, denom
    logical :: divisible_by

    divisible_by = .false.

    if( modulo(num, denom) .eq. 0 ) then
       divisible_by = .true.
    end if

  end function divisible_by

  subroutine simple_fizzbuzz(start, end)
    integer, intent(in) :: start, end
    integer :: iter

    iter = start

    do while(iter < end)

       if ( (divisible_by(iter,3)) .and. divisible_by(iter,5) ) then
          write(*,*) "Fizzbuzz"
       elseif ( divisible_by(iter,3) ) then
          write(*,*) "Fizz"
       elseif ( divisible_by(iter,5) ) then
          write(*,*) "Buzz"
       else
          write(*,'(I3)') iter
       end if
       iter = iter + 1
    enddo

  end subroutine simple_fizzbuzz

end module simple_fizzbuzz_mod

program fizzbuzz
  use simple_fizzbuzz_mod, only : simple_fizzbuzz
  integer, parameter :: start = 1
  integer, parameter ::  end = 101

  !call naive_fizzbuzz()
  call simple_fizzbuzz(start, end)
  
end program fizzbuzz

subroutine naive_fizzbuzz()
  integer :: i = 1
  
  do while(i < 16)
     if ( modulo(i,3).eq.0 .and. modulo(i,5).eq.0 ) then
        write(*,*) "Fizzbuzz"
     elseif ( modulo(i,3).eq.0 ) then
        write(*,*) "Fizz"
     elseif ( modulo(i,5).eq.0 ) then
        write(*,*) "Buzz"
     else
        write(*,'(I3)') i
     end if
     i = i + 1
  enddo
     
end subroutine naive_fizzbuzz

File: c/fizzbuzz.c
#include <stdio.h>
#include <string.h>

void simple_fizzbuzz();

int main() {
    void simple_fizzbuzz();
    void naive_fizzbuzz();
    void improved_naive_fizzbuzz();

    simple_fizzbuzz();
    printf("\n");
    naive_fizzbuzz();
    printf("\n");
    improved_naive_fizzbuzz();

    return 0;
}

void simple_fizzbuzz() {
    int i;

    for ( i = 1; i < 100; i++ ) {
        if ( ( i % 3 == 0 ) ) {
            printf("Fizz");
        }
        if ( ( i % 5 == 0) ) {
            printf("Buzz");
        }
        if ( !( i % 3 == 0 ) && !( i % 5 == 0) ) {
            printf( "%d", i);
        }
        printf(" ");
    }
}

void naive_fizzbuzz() {
    int i, mod3, mod5;

    for ( i = 1; i < 100; i++ ) {
        mod3 = ( i % 3 == 0 );
        mod5 = ( i % 5 == 0 );

        if ( mod3 ) {
            printf("Fizz");
        }
        if ( mod5 ) {
            printf("Buzz");
        }
        if ( !mod3 && !mod5 ) {
            printf("%d", i);
        }
        printf(" ");
    }
}

int divisible_by(int num, int denom) {
    return num % denom == 0;
}

void improved_naive_fizzbuzz() {

    int divisible_by(int, int);

    int i, mod3, mod5;

    for ( i = 1; i < 100; i++ ) {
        mod3 = divisible_by(i, 3);
        mod5 = divisible_by(i, 5);

        if ( mod3 ) {
            printf("Fizz");
        }
        if ( mod5 ) {
            printf("Buzz");
        }
        if ( !mod3 && !mod5 ) {
            printf("%d", i);
        }
        printf(" ");
    }
}